compiler/rustc_middle/src/ty/fold.rs

```diff
@@ -876,6 +877,77 @@ impl<'tcx> PredicateVisitor<'tcx> for HasEscapingVarsVisitor {
     }
 }
 
+crate struct CountBoundVars<'tcx> {
+    crate outer_index: ty::DebruijnIndex,
+    crate bound_tys: SsoHashSet<ty::BoundTy>,
+    crate bound_regions: SsoHashSet<ty::BoundRegion>,
+    crate bound_consts: SsoHashSet<ty::BoundVar>,
+    crate visited: SsoHashSet<Ty<'tcx>>,
+}
+
+impl<'tcx> CountBoundVars<'tcx> {
+    crate fn new() -> Self {
+        CountBoundVars {
+            outer_index: ty::INNERMOST,
+            bound_tys: SsoHashSet::default(),
+            bound_regions: SsoHashSet::default(),
+            bound_consts: SsoHashSet::default(),
+            visited: SsoHashSet::default(),
+        }
+    }
+}
+
+impl<'tcx> TypeVisitor<'tcx> for CountBoundVars<'tcx> {
+    fn visit_binder<T: TypeFoldable<'tcx>>(&mut self, t: &Binder<T>) -> bool {
+        self.outer_index.shift_in(1);
+        let result = t.super_visit_with(self);
+        self.outer_index.shift_out(1);
+        result
+    }
+
+    fn visit_ty(&mut self, t: Ty<'tcx>) -> bool {
+        if t.outer_exclusive_binder >= self.outer_index || !self.visited.insert(t) {
```

Comment:
isn't this incorrect?

Don't you instead want to exit early if `t.outer_exclusive_binder <= self.outer_index`?

Reply:
Ugh yeah